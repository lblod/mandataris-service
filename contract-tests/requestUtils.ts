let sessionUri = '';
import { v4 as uuidv4 } from 'uuid';
import { querySudo } from '@lblod/mu-auth-sudo';

export async function setSession(newSessionUri: string) {
  sessionUri = newSessionUri;
}

export async function getSession() {
  return sessionUri;
}

export async function mockLogin(groupUri, accountUri, roleName) {
  const id = uuidv4();
  const uri = `http://data.lblod.info/id/sessions/${id}`;
  const query = `              INSERT DATA {

                GRAPH <http://mu.semte.ch/graphs/sessions> {
                  <${uri}> <http://mu.semte.ch/vocabularies/session/account> <${accountUri}> .
                  <${uri}> <http://purl.org/dc/terms/modified> "${new Date().toISOString()}"^^<http://www.w3.org/2001/XMLSchema#dateTime> .
                  <${uri}> <http://mu.semte.ch/vocabularies/ext/sessionGroup> <${groupUri}> .
                  <${uri}> <http://mu.semte.ch/vocabularies/core/uuid> "${id}" .
                  <${uri}> <http://mu.semte.ch/vocabularies/ext/sessionRole> "${roleName}" .
                }
              }`;
  await querySudo(query);
  sessionUri = uri;
  return uri;
}

export async function userRequest(
  method: string,
  url: string,
  body: unknown = undefined,
  extraHeaders = {},
) {
  const response = await fetch(url, {
    method,
    headers: {
      'Content-Type': 'application/json',
      'mu-session-id': sessionUri,
      ...extraHeaders,
    },
    body: body ? JSON.stringify(body) : undefined,
  });
  const responseBody = await response.text();
  const headers = response.headers;
  const status = response.status;
  if (process.env.LOG_RESPONSES) {
    console.log(`Response from ${url}:
      status: ${status}
      body: ${body}`);
  }
  return {
    status,
    body: responseBody,
    headers,
  };
}
