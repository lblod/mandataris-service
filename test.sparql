  PREFIX mandaat: <http://data.vlaanderen.be/ns/mandaat#>
  PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
  PREFIX besluit: <http://data.vlaanderen.be/ns/besluit#>  
  PREFIX person: <http://www.w3.org/ns/person#>
  PREFIX mu: <http://mu.semte.ch/vocabularies/core/>
  PREFIX person: <http://www.w3.org/ns/person#>
  PREFIX persoon: <http://data.vlaanderen.be/ns/persoon#>
  PREFIX foaf: <http://xmlns.com/foaf/0.1/>
  PREFIX adms: <http://www.w3.org/ns/adms#>
  PREFIX skos: <http://www.w3.org/2004/02/skos/core#>

   INSERT DATA {
    GRAPH <http://mu.semte.ch/graphs/besluiten-consumed974816591f269bb7d74aa1720922651529f3d3b2a787f5c60b73e5a0384950a4/LoketLB-mandaatGebruiker> {
      <http://data.lblod.info/id/personen/1234567689> a person:Person;
          mu:uuid """1234567689""";
          persoon:gebruikteVoornaam """Jonas""";
          foaf:familyName """De tester""";
          persoon:geslacht <http://publications.europa.eu/resource/authority/human-sex/FEMALE>;
          foaf:name """JJ""";
          persoon:heeftGeboorte <http://data.lblod.info/id/geboortes/d9e8f8fa-8f5e-4d97-920e-70cff639de11>.
    }
  }

