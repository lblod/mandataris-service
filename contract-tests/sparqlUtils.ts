import { querySudo } from '@lblod/mu-auth-sudo';
import { expect } from '@jest/globals';
import { getSession } from './requestUtils';
const deltaPropagationTimeout = parseInt(
  process.env.DELTA_PROPAGATION_TIMEOUT || '200',
);

export async function expectQueryToHaveResults(
  query: string,
  extraHeaders = {},
) {
  const result = await querySudo(query, extraHeaders);
  expect(result.results.bindings.length).toBeGreaterThan(0);
}

export async function runSudoQuery(query: string, extraHeaders = {}) {
  const result = await querySudo(query, extraHeaders);
  return result.results.bindings;
}

export async function expectQueryToHaveNoResults(
  query: string,
  extraHeaders = {},
) {
  const result = await querySudo(query, extraHeaders);
  expect(result.results.bindings.length).toBe(0);
}

export async function expectUserQueryToHaveResults(query: string) {
  const result = await querySudo(query, {
    'mu-auth-sudo': 'false',
    'mu-session-id': getSession(),
  });
  expect(result.results.bindings.length).toBeGreaterThan(0);
}

export async function runUserQuery(query: string) {
  const result = await querySudo(query, {
    'mu-auth-sudo': 'false',
    'mu-session-id': getSession(),
  });
  return result.results.bindings;
}

export async function expectUserQueryToHaveNoResults(query: string) {
  const result = await querySudo(query, {
    'mu-auth-sudo': 'false',
    'mu-session-id': getSession(),
  });
  expect(result.results.bindings.length).toBe(0);
}

export async function getDeltas() {
  await new Promise((resolve) => setTimeout(resolve, deltaPropagationTimeout));
  return globalThis.deltas;
}

export async function clearDeltas() {
  await new Promise((resolve) => setTimeout(resolve, deltaPropagationTimeout));
  globalThis.deltas.splice(0, globalThis.deltas.length);
}
