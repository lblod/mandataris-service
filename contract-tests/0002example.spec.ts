import 'dotenv/config';

import { describe, expect, test, beforeAll } from '@jest/globals';
import { mockLogin, userRequest } from './requestUtils';
import { beforeEach } from 'node:test';
import { clearDeltas, getDeltas } from './sparqlUtils';

describe('index', () => {
  beforeAll(async () => {
    await mockLogin(
      'http://data.lblod.info/id/bestuurseenheden/5116efa8-e96e-46a2-aba6-c077e9056a96',
      'http://data.lblod.info/id/accounts/1234',
      'LoketLB-mandaatGebruiker',
    );
  });
  beforeEach(async () => {
    await clearDeltas();
  });

  test('hello world after', async () => {
    const { body } = await userRequest(
      'GET',
      'http://target/mandataris-api/organen/0921318d-c195-4fbf-aadf-2534cdd63c7d/activeMembers',
    );
    expect(body).toMatchSnapshot();
    expect(await getDeltas()).toHaveLength(0);
  });
});
