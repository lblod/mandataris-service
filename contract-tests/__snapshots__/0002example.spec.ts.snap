// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index hello world after 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /mandataris-api/organen/0921318d-c195-4fbf-aadf-2534cdd63c7d/activeMembers</pre>
</body>
</html>
"
`;
